octopusdeploy_framework/datasource_tenant_projects.go

```diff
@@ -0,0 +1,201 @@
+			// todo: should consider using concurrency
+			tenantsByProjectID, err := getTenantByProjectID(t.Client, projectID, spaceID)
+			if err != nil {
+				resp.Diagnostics.AddError("unable to load tenant data for project", err.Error())
+				return
+			}
+			tenantData = append(tenantData, tenantsByProjectID...)
+		}
+	} else {
+		tenantQuery := tenants.TenantsQuery{
+			IDs: tenantIDs,
+		}
+		tenantResource, err := tenants.Get(t.Client, spaceID, tenantQuery)
+		if err != nil {
+			resp.Diagnostics.AddError("unable to load tenant data", err.Error())
+			return
+		}
+		tenantData, err = tenantResource.GetAllPages(t.Client.Sling())
+		if err != nil {
+			resp.Diagnostics.AddError("unable to load tenant data", err.Error())
+			return
+		}
+	}
+
+	// todo: refactor
+	flattenedTenantProjects := make([]any, 0)
+	for _, tenant := range tenantData {
+		if projectIDs == nil {
+			for projectID, envIDs := range tenant.ProjectEnvironments {
+				if environmentIDs != nil {
+					for _, envID := range envIDs {
+						if slices.Contains(environmentIDs, envID) {
+							flattenedTenantProjects = append(flattenedTenantProjects, flattenTenantProject(tenant, projectID))
+							break
+						}
+					}
+				} else {
+					flattenedTenantProjects = append(flattenedTenantProjects, flattenTenantProject(tenant, projectID))
+				}
+			}
+			continue
+		}
+		for _, projectID := range projectIDs {
+			if envIDs, ok := tenant.ProjectEnvironments[projectID]; ok {
+				if environmentIDs != nil {
+					for _, envID := range envIDs {
+						if slices.Contains(environmentIDs, envID) {
+							flattenedTenantProjects = append(flattenedTenantProjects, flattenTenantProject(tenant, projectID))
+							break
+						}
+					}
+				} else {
+					flattenedTenantProjects = append(flattenedTenantProjects, flattenTenantProject(tenant, projectID))
+				}
+			}
+
+		}
+	}
+
+	data.TenantProjects, _ = types.ListValueFrom(ctx, types.ObjectType{AttrTypes: tenantProjectType()}, flattenedTenantProjects)
+	resp.Diagnostics.Append(resp.State.Set(ctx, &data)...)
+}
+
+func getTenantByProjectID(c *client.Client, projectID string, spaceID string) ([]*tenants.Tenant, error) {
+	tenantQuery := tenants.TenantsQuery{
+		ProjectID: projectID,
+	}
+	tenantResource, err := tenants.Get(c, spaceID, tenantQuery)
+	if err != nil {
+		return nil, err
+	}
+	tenantData, err := tenantResource.GetAllPages(c.Sling())
+	if err != nil {
+		return nil, err
+	}
+	return tenantData, nil
+}
+
+func buildTenantProjectID(spaceID string, projectID string, tenantID string) string {
+	return fmt.Sprintf("%s:%s:%s", spaceID, projectID, tenantID)
+}
+
+func tenantProjectType() map[string]attr.Type {
+	return map[string]attr.Type{
+		"id":              types.StringType,
+		"tenant_id":       types.StringType,
+		"project_id":      types.StringType,
+		"environment_ids": types.ListType{ElemType: types.StringType},
+	}
+}
+
+func flattenTenantProject(tenant *tenants.Tenant, projectID string) attr.Value {
```

Comment:
this is a personal preference: not a fan of the flatten/expand terminology. I would love to see this dropped as we migrate, but that is just a personal thing. Expand and flatten don't actually mean anything, as far as I am concerned.

Reply:
agreed, its more of a conversion here.